web/concrete/src/Attribute/Key/CollectionKey.php

```diff
@@ -38,104 +41,122 @@ public static function getAttributes($cID, $cvID, $method = 'getValue') {
-			'cID' => $nvc->getCollectionID(),
-			'cvID' => $nvc->getVersionID(),
-			'akID' => $this->getAttributeKeyID(),
-			'avID' => $av->getAttributeValueID()
-		), array('cID', 'cvID', 'akID'));
-		unset($av);
-	}
-
-	public static function add($at, $args, $pkg = false) {
-
-		// legacy check
-		$fargs = func_get_args();
-		if (count($fargs) >= 5) {
-			$at = $fargs[4];
-			$pkg = false;
-			$args = array('akHandle' => $fargs[0], 'akName' => $fargs[1], 'akIsSearchable' => $fargs[2]);
-		}
-
-		CacheLocal::delete('collection_attribute_key_by_handle', $args['akHandle']);
-
-		$ak = parent::add('collection', $at, $args, $pkg);
-		return $ak;
-	}
-
-	public function delete() {
-		parent::delete();
-		$db = Loader::db();
-		$r = $db->Execute('select avID from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-		while ($row = $r->FetchRow()) {
-			$db->Execute('delete from AttributeValues where avID = ?', array($row['avID']));
-		}
-		$db->Execute('delete from CollectionAttributeValues where akID = ?', array($this->getAttributeKeyID()));
-	}
+    public static function getColumnHeaderList()
+    {
+        return parent::getList('collection', array('akIsColumnHeader' => 1));
+    }
+
+    public static function getSearchableIndexedList()
+    {
+        return parent::getList('collection', array('akIsSearchableIndexed' => 1));
+    }
+
+    public static function getSearchableList()
+    {
+        return parent::getList('collection', array('akIsSearchable' => 1));
+    }
+
+    public function getAttributeValue($avID, $method = 'getValue')
+    {
+        $av = CollectionAttributeValue::getByID($avID);
+        if (is_object($av)) {
+            $av->setAttributeKey($this);
+            $value = $av->{$method}();
+            $av->__destruct();
+            unset($av);
+            return $value;
+        }
+    }
+
+    public static function getByID($akID)
+    {
+        $ak = new static();
+        $ak->load($akID);
+        if ($ak->getAttributeKeyID() > 0) {
+            return $ak;
+        }
+    }
+
+    public static function getByHandle($akHandle)
+    {
+        $ak = CacheLocal::getEntry('collection_attribute_key_by_handle', $akHandle);
+        if (is_object($ak)) {
+            return $ak;
+        } else if ($ak == -1) {
+            return false;
+        }
+
+        $ak = new static();
+        $ak->load($akHandle, 'akHandle');
+        if ($ak->getAttributeKeyID() < 1) {
+            $ak = -1;
+        }
+
+        CacheLocal::set('collection_attribute_key_by_handle', $akHandle, $ak);
+
+        if ($ak === -1) {
+            return false;
+        }
+
+        return $ak;
+    }
+
+    /**
+     * Get list of attributes for collection category
+     * @param null $akCategoryHandle Ignored; included for legacy support
+     * @param null $filters Ignored; included for legacy support
+     * @return array
+     */
+    public static function getList($akCategoryHandle = null, $filters = null)
```

Comment:
@KorvinSzanto are you saying to go with the `func_get_args()`? or something else,

for the `$akCategoryHandle` we can do some fancy stuff with `get_called_class()`

For example, with the add() method in this class, it has the signature of:
`protected static function add($akCategoryHandle, $type, $args, $pkg = false)`

everything else that extends it has the signature:
`protected static function add($type, $args, $pkg = false)`

which then calls `parent::add('page', $type, $args, $pkg);`

The first parameter can be resolved from the class name, eg UserKey -> user, PageKey -> page, etc

Below is some code that is backwards compatible, and works, it removes the need for the `$akCategoryHandle` parameter by magically figuring it out

```
        $txt = Core::make('helper/text');

        $fargs = func_get_args();
        if (count($fargs) >= 4) {
            $akCategoryHandle = $fargs[0];
            $type = $fargs[1];
            $args = $fargs[2];
            $pkg = $fargs[3];
        } else {
            $class = get_called_class();
            $last = array_pop(explode('\\', $class));
            $akCategoryHandle = $txt->uncamelcase(substr($last, 0, -3)); //strip off `Key`
        }
```

Thoughts?